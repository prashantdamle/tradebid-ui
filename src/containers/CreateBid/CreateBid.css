.CreateBid {
  padding: 40px;
  margin: 0 auto;
}

.CreateBid h2 {
  margin-bottom: 0;
}

.CreateBid small {
  display: block;
}
