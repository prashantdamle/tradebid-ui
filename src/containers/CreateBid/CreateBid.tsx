import * as React from "react";
import { Button, Form, FormGroup, Input, Label, Spinner } from "reactstrap";
import { Bid, HoursType } from "../../types/model";
import { ErrorMessages, RouteURLs } from "../../types/enum";
import { Link, useNavigate, useSearchParams } from "react-router-dom";
import { UserContext } from "../../provider/UserProvider";
import useAxios from "axios-hooks";
import { BASE_URL } from "../../constants";

import "./CreateBid.css";

interface Props {}

const CreateBid = (props: Props) => {
  const navigate = useNavigate();
  const { user } = React.useContext(UserContext);
  const userId = user?.userId;
  const [searchParams] = useSearchParams();
  const pid = searchParams.get("pid");
  const pname = searchParams.get("pname");

  const [bidTypeValue, setBidTypeValue] = React.useState<HoursType>(
    HoursType.FIXED
  );
  const [priceValue, setPriceValue] = React.useState("");
  const [errorMsg, setErrorMsg] = React.useState("");

  const [{ data, loading, error }, execute] = useAxios<Bid>(
    {
      method: "post",
      url: `${BASE_URL}/user/${userId}/customer-project/${pid}/bid`,
    },
    { manual: true }
  );

  const submitHandler = (e: React.FormEvent<HTMLFormElement>) => {
    e?.preventDefault();

    if (!bidTypeValue || !priceValue) {
      setErrorMsg(ErrorMessages.FORM_ERROR);
      return;
    }

    const requestData = {
      userId,
      bidType: bidTypeValue,
      price: Number(priceValue),
    };

    execute({ data: requestData });
  };

  React.useEffect(() => {
    if (data) {
      data?.bidId
        ? navigate(RouteURLs.TRADESPERSON_DASHBOARD)
        : setErrorMsg(ErrorMessages.GENERIC_ERROR);
    }
  }, [data, navigate]);

  React.useEffect(() => {
    if (error) setErrorMsg(ErrorMessages.GENERIC_ERROR);
  }, [error]);

  return (
    <div className="CreateBid">
      <Link to={RouteURLs.TRADESPERSON_DASHBOARD}>Back to Dashboard</Link>
      <br />
      <br />
      <h2>{`Create a bid for project '${pname}'`}</h2>
      <small>(all fields mandatory)</small>
      <br />
      <Form onSubmit={submitHandler}>
        <FormGroup floating>
          <Input
            id="bidType"
            name="select"
            type="select"
            value={bidTypeValue}
            onChange={(e) => setBidTypeValue(e.target.value as HoursType)}
          >
            {Object.values(HoursType).map((hourType) => (
              <option key={hourType}>{hourType}</option>
            ))}
          </Input>
          <Label for="bidType">Select</Label>
        </FormGroup>{" "}
        <FormGroup floating>
          <Input
            id="price"
            name="price"
            placeholder="Price"
            type="tel"
            value={priceValue}
            onChange={(e) => setPriceValue(e.target.value)}
          />
          <Label for="price">Price</Label>
        </FormGroup>{" "}
        <br />
        <Button type="submit" disabled={loading}>
          <span>{loading ? "Loading" : "Submit"} </span>
          {loading && <Spinner size="sm">Loading...</Spinner>}
        </Button>
        {errorMsg && <small className="Error">{errorMsg}</small>}
      </Form>
    </div>
  );
};

export default CreateBid;
