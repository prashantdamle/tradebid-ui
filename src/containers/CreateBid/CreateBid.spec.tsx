import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CreateBid from "./CreateBid";

const mockedUsedNavigate = jest.fn();
const searchParams = {
  pid: "123",
  pname: "project test name",
  get: (pid: string) => {
    return pid;
  },
};

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
  useSearchParams: () => [searchParams],
}));

it("renders correctly", () => {
  const { container } = render(
    <MemoryRouter>
      <CreateBid />
    </MemoryRouter>
  );

  expect(container).toMatchSnapshot();
});
