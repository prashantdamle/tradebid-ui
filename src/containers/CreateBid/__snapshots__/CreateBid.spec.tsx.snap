// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    class="CreateBid"
  >
    <a
      href="/tradesperson-dashboard"
    >
      Back to Dashboard
    </a>
    <br />
    <br />
    <h2>
      Create a bid for project 'pname'
    </h2>
    <small>
      (all fields mandatory)
    </small>
    <br />
    <form
      class=""
    >
      <div
        class="mb-3 form-floating"
      >
        <select
          class="form-select"
          id="bidType"
          name="select"
        >
          <option>
            HOURLY
          </option>
          <option>
            FIXED
          </option>
        </select>
        <label
          class="form-label"
          for="bidType"
        >
          Select
        </label>
      </div>
       
      <div
        class="mb-3 form-floating"
      >
        <input
          class="form-control"
          id="price"
          name="price"
          placeholder="Price"
          type="tel"
          value=""
        />
        <label
          class="form-label"
          for="price"
        >
          Price
        </label>
      </div>
       
      <br />
      <button
        class="btn btn-secondary"
        type="submit"
      >
        <span>
          Submit
           
        </span>
      </button>
    </form>
  </div>
</div>
`;
